File: leetcode/binary_tree_right_side_view.js
// https://leetcode.com/problems/binary-tree-right-side-view/
// 199. Binary Tree Right Side View
/**
 * @param {TreeNode} root
 * @return {number[]}
 */

var rightSideView = function (root) {
    let map = {};
    traverse(root, 0);
    let keys = Object.keys(map);
    let res = new Array(keys.length);
    for (let i = 0; i < keys.length; i++) {
        res[keys[i]] = map[keys[i]]
    }
    function traverse(root, depth) {
        if (!root) {
            return;
        }
        traverse(root.left, depth + 1);
        traverse(root.right, depth + 1);
        map[depth] = root.val;
    }

    return res;
};

function main() {
    let input =[1,2,3,4,5,6,null,8,9,10,null,12,13,14,15]
    let root = makeTree(input);
    let res = rightSideView(root);
    console.log(res);

}

function TreeNode(val, left, right) {
    this.val = (val === undefined ? 0 : val)
    this.left = (left === undefined ? null : left)
    this.right = (right === undefined ? null : right)
}
function makeTree(input) {
    function getTree(subRootIndex) {
        if (subRootIndex >= input.length || input[subRootIndex] == null) {
            return null
        }
        return new TreeNode(input[subRootIndex], getTree((subRootIndex * 2) + 1), getTree((subRootIndex + 1) * 2))
    }
    return getTree(0)
}

main();

File: leetcode/binary_tree_right_side_view.py
# Definition for a binary tree node.
from typing import List, Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


def getTree(nodes: List[int]) -> TreeNode:
    def recurse(index):
        if index >= len(nodes):
            return None
        if nodes[index] is None:
            return None
        subroot = TreeNode(nodes[index])
        left = recurse((index * 2) + 1)
        right = recurse((index * 2) + 2)
        subroot.left = left
        subroot.right = right
        return subroot

    return recurse(0)


class Solution:
    def rightSideView(self, root: Optional[TreeNode]) -> List[int]:
        rightView = []

        def recurse(subroot: TreeNode, level: int):
            if subroot is None:
                return
            if len(rightView) <= level:
                rightView.append(subroot.val)
            recurse(subroot.right, level + 1)
            recurse(subroot.left, level + 1)
            return

        recurse(root, 0)
        return rightView


if __name__ == '__main__':
    list = []

    tree = getTree(list)
    res = Solution().rightSideView(tree)
    print(res)
